sysfs-val.c

```diff
@@ -0,0 +1,242 @@
+ * along with mce-plugin-libhybris; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "sysfs-val.h"
+
+#include "plugin-logging.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <fcntl.h>
+
+/* ========================================================================= *
+ * TYPES
+ * ========================================================================= */
+
+struct sysfsval_t
+{
+    char *sv_path;
+    int   sv_file;
+    int   sv_curr;
+};
+
+/* ========================================================================= *
+ * PROTOS
+ * ========================================================================= */
+
+static void        sysfsval_ctor      (sysfsval_t *self);
+static void        sysfsval_dtor      (sysfsval_t *self);
+
+sysfsval_t        *sysfsval_create    (void);
+void               sysfsval_delete    (sysfsval_t *self);
+bool               sysfsval_open      (sysfsval_t *self, const char *path);
+void               sysfsval_close     (sysfsval_t *self);
+const char        *sysfsval_path      (const sysfsval_t *self);
+int                sysfsval_get       (const sysfsval_t *self);
+bool               sysfsval_set       (sysfsval_t *self, int value);
+void               sysfsval_invalidate(sysfsval_t *self);
+bool               sysfsval_refresh   (sysfsval_t *self);
+
+/* ========================================================================= *
+ * CODE
+ * ========================================================================= */
+
+static void
+sysfsval_ctor(sysfsval_t *self)
+{
+    self->sv_path = 0;
+    self->sv_file = -1;
+    self->sv_curr = -1;
+}
+
+static void
+sysfsval_dtor(sysfsval_t *self)
+{
+    sysfsval_close(self);
+}
+
+sysfsval_t *
+sysfsval_create(void)
+{
+    sysfsval_t *self = calloc(1, sizeof *self);
+    sysfsval_ctor(self);
+    return self;
+}
+void
+sysfsval_delete(sysfsval_t *self)
+{
+    if( self )  {
+        sysfsval_dtor(self);
+        free(self);
+    }
+}
+
+bool
+sysfsval_open(sysfsval_t *self, const char *path)
+{
+    bool ack = false;
+
+    sysfsval_close(self);
+
+    if( !path )
+        goto EXIT;
+
+    if( (self->sv_path = strdup(path)) == 0 )
+        goto EXIT;
+
+    if( (self->sv_file = open(path, O_RDWR)) == -1 )
+        goto EXIT;
+
+    mce_log(LOG_DEBUG, "%s: opened", sysfsval_path(self));
+
+    /* Note: Current value is not fetched by default */
+
+    ack = true;
+
+EXIT:
+    if( !ack )
+        sysfsval_close(self);
+
+    return ack;
+}
+
+void
+sysfsval_close(sysfsval_t *self)
+{
+    if( self->sv_file != -1 ) {
+        mce_log(LOG_DEBUG, "%s: closed", sysfsval_path(self));
+        close(self->sv_file), self->sv_file = -1;
+    }
+
+    free(self->sv_path), self->sv_path = 0;
+}
+
+const char *
+sysfsval_path(const sysfsval_t *self)
+{
+    return self->sv_path ?: "unknown";
```

Comment:
nitpick: When a string is returned for null path it isn't possible to easily distinguish "error" case, if needed.. Also, it isn't possible to open a sysfsval without a path anyway..